[WebGPU] nullptr crash in index buffer validation

mwyrzykowski opened the pull request:
#### 07dade918043f99fb9ff8d13cbba3414b85a054d
<pre>
[WebGPU] nullptr crash in index buffer validation
<a href="https://bugs.webkit.org/show_bug.cgi?id=291545">https://bugs.webkit.org/show_bug.cgi?id=291545</a>
<a href="https://rdar.apple.com/149246187">rdar://149246187</a>

Reviewed by Cameron McCormack.

293608@main moved a RefPtr to a function callback inside a for-loop,
so upon the second iteration of the loop the RefPtr was nullptr and we
crashed with a nullptr access.

Address this by looking up the buffer in the for loop.

* LayoutTests/fast/webgpu/nocrash/fuzz-291545-expected.txt: Added.
* LayoutTests/fast/webgpu/nocrash/fuzz-291545.html: Added.
Add regression test.

* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::CommandEncoder):

Canonical link: <a href="https://commits.webkit.org/293702@main">https://commits.webkit.org/293702@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/67207f2314389c182988019ab9885f70dc439236

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99647 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19297 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/9558 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104777 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50243 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/19586 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27730 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75851 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32950 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102654 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14914 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89977 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56210 "Found 1 new API test failure: /WPE/TestCookieManager:/webkit/WebKitCookieManager/replace-get-all-cookies (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14714 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7963 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/49605 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/84677 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8048 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107136 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26762 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/19536 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84809 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-on-non-root.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27126 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86181 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84328 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21411 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29003 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6710 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/20552 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/26702 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/31904 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26517 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29833 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28087 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebGPU/WebGPU/CommandEncoder.mm:
```diff
@@ -161,8 +161,9 @@ static MTLStoreAction storeAction(WGPUStoreOp storeOp, bool hasResolveTarget = f
-                    commandBuffer.addPostCommitHandler([apiBuffer = WTFMove(apiBuffer)](id<MTLCommandBuffer>) {
-                        apiBuffer->clearMustTakeSlowIndexValidationPath();
+                    commandBuffer.addPostCommitHandler([bufferIdentifier, device = Ref { commandBuffer.device() }](id<MTLCommandBuffer>) {
+                        if (RefPtr apiBuffer = device->lookupBuffer(bufferIdentifier))
```
Is there any expected condition where this would return null? Should we assert this, rather than null check it? Or could something in between `addPostCommitHandler` being called and the callback being run cause the buffer to go away?

mwyrzykowski replied:
CommandEncoder will retain all the BindGroups in a CommandBuffer which retains the Buffer instances too.

However because in:
```
void CommandBuffer::makeInvalidDueToCommit(NSString* lastError)
{
...
    [m_commandBuffer addCompletedHandler:[protectedThis = Ref { *this }](id<MTLCommandBuffer>) {
        protectedThis->m_commandBufferComplete.signal();
        protectedThis->m_device->protectedQueue()->scheduleWork([protectedThis = WTFMove(protectedThis)]() mutable {
            protectedThis->m_cachedCommandBuffer = nil;
            protectedThis->m_commandEncoder = nullptr;
```
`m_commandEncoder = nullptr;`, then once that occurs, nothing may be retaining the Buffer, so it could get so `device->lookupBuffer(bufferIdentifier)` could retain nullptr.

The alternative would be to retain `apiBuffer` in the post-commit handler, but this should be equivalent without the additional retain.

mwyrzykowski committed:
[WebGPU] nullptr crash in index buffer validation
https://bugs.webkit.org/show_bug.cgi?id=291545
rdar://149246187

Reviewed by Cameron McCormack.

293608@main moved a RefPtr to a function callback inside a for-loop,
so upon the second iteration of the loop the RefPtr was nullptr and we
crashed with a nullptr access.

Address this by looking up the buffer in the for loop.

* LayoutTests/fast/webgpu/nocrash/fuzz-291545-expected.txt: Added.
* LayoutTests/fast/webgpu/nocrash/fuzz-291545.html: Added.
Add regression test.

* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::CommandEncoder):

Canonical link: https://commits.webkit.org/293702@main
